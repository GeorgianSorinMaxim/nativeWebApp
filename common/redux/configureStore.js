// @flow

import { createStore, applyMiddleware } from 'redux';

import { createEpicMiddleware } from 'redux-observable';
import { composeWithDevTools } from 'redux-devtools-extension';

import { rootReducer, rootEpic } from './root';

const epicMiddleware = store =>
  createEpicMiddleware(rootEpic, { dependecies: { dispatch: store.dispatch } })(
    store
  );

const getEnhancers = () => {
  const enhancers = applyMiddleware(epicMiddleware);

  /* global __DEV__ */
  if (__DEV__) {
    /* eslint-disable */
    const devToolsCompose = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__
      ? window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__
      : composeWithDevTools;
    /* eslint-enable */

    return devToolsCompose({
      realtime: true,
      port: 8010
    })(enhancers);
  }

  return enhancers;
};

export default function configureStore() {
  return createStore(rootReducer, getEnhancers());
}
