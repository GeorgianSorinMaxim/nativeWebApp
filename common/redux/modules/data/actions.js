// @flow

export const APP_LAUNCHED = 'data/APP_LAUNCHED';

export const START_FETCH_DATA = 'data/START_FETCH';

export const DATA_FETCH_SUCCESS = 'data/FETCH_SUCCESS';
export const DATA_FETCH_FAILURE = 'data/FETCH_FAILURE';

type AppLaunched = {
  type: 'data/APP_LAUNCHED'
};

type StartDataFetch = {
  type: 'data/START_FETCH'
};

type DataFetchSuccess = {
  type: 'data/FETCH_SUCCESS',
  payload: Object
};

type DataFetchFailure = {
  type: 'data/FETCH_FAILURE',
  payload: Object
};

export function appLaunched(): AppLaunched {
  return { type: APP_LAUNCHED };
}

export function startFetchingData(): StartDataFetch {
  return { type: START_FETCH_DATA };
}

export function dataFetchSuccess(payload: Object): DataFetchSuccess {
  return { type: DATA_FETCH_SUCCESS, payload };
}

export function dataFetchFailure(error: Object): DataFetchFailure {
  return { type: DATA_FETCH_FAILURE, payload: { error } };
}

export type Action =
  | AppLaunched
  | StartDataFetch
  | DataFetchSuccess
  | DataFetchFailure;
