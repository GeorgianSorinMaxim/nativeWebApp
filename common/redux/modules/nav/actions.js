// @flow
import { NavigationActions } from 'react-navigation';

const navToPath = path => NavigationActions.navigate({ routeName: path });
// const navToPathWithParams = (path: string, params: Object) =>
//   NavigationActions.navigate({ routeName: path, params });

const navToHome = () => navToPath('Home');
const navToFeature = () => navToPath('Feature');

export const actions = {
  navToHome,
  navToFeature
};
