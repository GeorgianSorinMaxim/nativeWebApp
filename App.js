import 'babel-polyfill';
import React, { Component } from 'react';
import { Provider } from 'react-redux';

import type { Store } from 'redux';

import configureStore from './common/redux/configureStore';
import * as actions from './common/redux/modules/data/actions';

import HomeScreen from './common/components/HomeScreen';

type Props = {};
type State = {
  loading: boolean
};

export class App extends React.Component<Props, State> {
  store: Store<any, any>;

  state = {
    loading: true
  };

  componentDidMount() {
    this.asyncComponentDidMount();
  }

  asyncComponentDidMount = async () => {
    this.store = configureStore();

    this.setState({ loading: false });

    this.store.dispatch(actions.appLaunched());
  };

  renderLoadedApp() {
    return (
      <Provider store={this.store}>
        <HomeScreen />
      </Provider>
    );
  }

  render() {
    return this.state.loading ? null : this.renderLoadedApp();
  }
}

export default App;
