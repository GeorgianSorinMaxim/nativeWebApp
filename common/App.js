// @flow

import 'babel-polyfill';
import React from 'react';
import { Provider } from 'react-redux';

import type { Store } from 'redux';

import { navMiddleware } from './redux/NavMiddleware';
import configureStore from './redux/configureStore';
import * as actions from './redux/modules/data/actions';
import NavigationContainer from './components/NavigationContainer';

type Props = {};
type State = {
  loading: boolean
};

export class App extends React.Component<Props, State> {
  store: Store<any, any>;

  state = {
    loading: true
  };

  componentDidMount() {
    this.asyncComponentDidMount();
  }

  asyncComponentDidMount = async () => {
    this.store = configureStore();

    this.setState({ loading: false });

    this.store.dispatch(actions.appLaunched());
  };

  renderLoadedApp() {
    return (
      <Provider store={this.store}>
        <NavigationContainer />
      </Provider>
    );
  }

  render() {
    return this.state.loading ? null : this.renderLoadedApp();
  }
}

export default App;
