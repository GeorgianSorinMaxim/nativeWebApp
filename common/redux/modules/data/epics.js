// @flow

import 'rxjs/add/operator/switchMap';
import 'rxjs/add/operator/mapTo';
import 'rxjs/add/operator/map';
import 'rxjs/add/operator/filter';
import 'rxjs/add/operator/catch';
import 'rxjs/add/operator/pluck';
import 'rxjs/add/observable/of';
import 'rxjs/add/observable/throw';
import 'rxjs/add/observable/fromPromise';
import 'rxjs/add/observable/from';

import { Observable } from 'rxjs/Observable';
import { combineEpics } from 'redux-observable';

import * as actions from './actions';

const startFetchDataEpic = action$ =>
  action$.ofType(actions.APP_LAUNCHED).mapTo(actions.startFetchingData());

const fetchDataEpic = action$ =>
  action$.ofType(actions.START_FETCH_DATA).switchMap(() =>
    fetch('https://hacker-news.firebaseio.com/v0/topstories.json?print=pretty')
      .then(
        response =>
          response.ok ? response.json() : actions.dataFetchFailure(FETCH_ERROR)
      )
      .then(
        result =>
          result.type === 'REQUEST_FAILED'
            ? result
            : actions.dataFetchSuccess(result)
      )
      .catch(() => actions.dataFetchFailure(FETCH_ERROR))
  );

export default combineEpics(startFetchDataEpic, fetchDataEpic);
