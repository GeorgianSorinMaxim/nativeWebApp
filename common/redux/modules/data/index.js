// @flow

import * as actions from './actions';

export { actions };
export { default as epics } from './epics';
export { default } from './reducer';

export type { Action } from './actions';
export type { State } from './reducer';
