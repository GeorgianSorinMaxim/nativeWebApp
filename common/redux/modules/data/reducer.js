// @flow

import * as actions from './actions';

import type { Action } from './actions';

export type State = {
  entities: Object,
  loading: boolean,
  error?: any
};

export const initialState: State = {
  entities: {},
  loading: false
};

export default function reducer(
  state: State = initialState,
  action: Action
): State {
  switch (action.type) {
    case actions.START_FETCH_DATA:
      return {
        ...state,
        loading: true
      };

    case actions.DATA_FETCH_SUCCESS:
      return {
        ...state,
        entities: action.payload,
        loading: false
      };

    case actions.DATA_FETCH_FAILURE:
      return {
        ...state,
        error: action.payload,
        loading: false
      };

    default:
      return state;
  }
}
