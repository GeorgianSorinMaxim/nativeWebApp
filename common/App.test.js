import React from 'react';

import configureStore from './redux/configureStore';
import * as actions from './redux/modules/data/actions';

import render from './utils/testUtils';
import App from './App';

const mockStore = {
  dispatch: jest.fn(),
  subscribe: jest.fn(),
  getState: () => ({
    data: {
      entities: [],
      loading: false
    }
  })
};

jest.mock('./redux/configureStore', () => jest.fn(() => mockStore));

describe('App', () => {
  describe('render', () => {
    it('renders nothing before mounting and initialising', () => {
      const { tree } = render(<App />);

      expect(tree).toMatchSnapshot();
    });
  });

  describe('asyncComponentDidMount', () => {
    it('initialises correctly', async () => {
      const { instance } = render(<App />);

      await instance.asyncComponentDidMount();
      const tree = instance.render();
      expect(tree).toMatchSnapshot();
      expect(instance.state.loading).toBe(false);
      expect(configureStore).toHaveBeenCalled();
      expect(mockStore.dispatch).toHaveBeenCalledWith(actions.appLaunched());
    });
  });
});
