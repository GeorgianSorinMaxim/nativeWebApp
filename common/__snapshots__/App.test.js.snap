// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App asyncComponentDidMount initialises correctly 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(HomeScreenContainer) />
</Provider>
`;

exports[`App render renders nothing before mounting and initialising 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#F5FCFF",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <View
    style={
      Object {
        "flex": 1,
        "marginVertical": 40,
        "width": "100%",
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "fontSize": 20,
          "margin": 10,
          "textAlign": "center",
        }
      }
    >
      Hello 90poe!
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "color": undefined,
          "marginBottom": 5,
          "textAlign": "center",
        }
      }
    >
      This is the ðŸ“± app
    </Text>
    <RCTScrollView
      data={Array []}
      disableVirtualization={false}
      getItem={[Function]}
      getItemCount={[Function]}
      horizontal={false}
      initialNumToRender={25}
      invertStickyHeaders={undefined}
      keyExtractor={[Function]}
      maxToRenderPerBatch={10}
      onContentSizeChange={[Function]}
      onEndReachedThreshold={2}
      onLayout={[Function]}
      onMomentumScrollEnd={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      renderItem={[Function]}
      scrollEventThrottle={50}
      stickyHeaderIndices={Array []}
      style={
        Object {
          "marginVertical": 20,
        }
      }
      updateCellsBatchingPeriod={50}
      windowSize={21}
    >
      <View />
    </RCTScrollView>
  </View>
</View>
`;
